File: Spark/src/main/java/lxpsee/top/spark/core/study/javaVerson/study4core/SortWordCountByValue.java
package lxpsee.top.spark.core.study.javaVerson.study4core;

import lxpsee.top.spark.core.study.constant.LPConstants;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;

/**
 * The world always makes way for the dreamer
 * Created by 努力常态化 on 2019/3/27 16:22.
 * <p>
 * 1、对文本文件内的每个单词都统计出其出现的次数。
 * 2、按照每个单词出现次数的数量，降序排序
 */
public class SortWordCountByValue {
    public static void main(String[] args) {
        JavaSparkContext sparkContext = new JavaSparkContext(new SparkConf().setMaster("local").setAppName("SortWordCountByValueJava"));
        JavaRDD<String> lines = sparkContext.textFile(LPConstants.SPARK_TXT_LOCAL_FILE_PATH);
        JavaRDD<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
            public Iterator<String> call(String s) throws Exception {
                return Arrays.asList(s.split(" ")).iterator();
            }
        });
        JavaPairRDD<String, Integer> pairs = words.mapToPair(new PairFunction<String, String, Integer>() {
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        });
        JavaPairRDD<String, Integer> wordCount = pairs.reduceByKey(new Function2<Integer, Integer, Integer>() {
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        });
        JavaPairRDD<Integer, String> valuesWC = wordCount.mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
            public Tuple2<Integer, String> call(Tuple2<String, Integer> tuple2) throws Exception {
                return new Tuple2<Integer, String>(tuple2._2, tuple2._1);
            }
        });
        JavaPairRDD<Integer, String> sortRDD = valuesWC.sortByKey(false);
        JavaPairRDD<String, Integer> sortByValue = sortRDD.mapToPair(new PairFunction<Tuple2<Integer, String>, String, Integer>() {
            public Tuple2<String, Integer> call(Tuple2<Integer, String> t) throws Exception {
                return new Tuple2<String, Integer>(t._2, t._1);
            }
        });


        sortByValue.foreach(new VoidFunction<Tuple2<String, Integer>>() {
            public void call(Tuple2<String, Integer> tuple2) throws Exception {
                System.out.println(tuple2._1 + " 出现 " + tuple2._2 + " 次");
            }
        });

        sparkContext.close();
    }
}

File: Spark/src/main/java/lxpsee/top/spark/core/study/scalaVerson/study4core/SortWordCountByValue.scala
package lxpsee.top.spark.core.study.scalaVerson.study4core

import lxpsee.top.spark.core.study.constant.LPConstants
import org.apache.spark.{SparkConf, SparkContext}

/**
  * The world always makes way for the dreamer
  * Created by 努力常态化 on 2019/3/27 16:32.
  */
object SortWordCountByValue {
  def main(args: Array[String]): Unit = {
    val context = new SparkContext(new SparkConf().setAppName("SortWordCountByValue").setMaster("local"))
    context.textFile(LPConstants.SPARK_TXT_LOCAL_FILE_PATH)
      .flatMap(_.split(" "))
      .map((_, 1))
      .reduceByKey(_ + _)
      .map(reduceRDD => (reduceRDD._2, reduceRDD._1))
      .sortByKey(false)
      .map(sortRDD => (sortRDD._2, sortRDD._1))
      .foreach(valueRDD => println(valueRDD._1 + " 出现 " + valueRDD._2))
  }
}
